core/domain/question_jobs_one_off.py

```diff
@@ -140,3 +142,105 @@ def reduce(key, values):
                 sum(ast.literal_eval(v) for v in values))])
         else:
             yield (key, values)
+
+
+class MissingQuestionMigrationOneOffJob(jobs.BaseMapReduceOneOffJobManager):
+    """This job is used to delete question commit log models for which question
+    models are missing.
+    """
+
+    @classmethod
+    def entity_classes_to_map_over(cls):
+        return [question_models.QuestionCommitLogEntryModel]
+
+    @staticmethod
+    def map(item):
+        if item.deleted:
+            return
+
+        question = question_services.get_question_by_id(
+            item.question_id, strict=False)
+        if question is None:
+            item.delete()
+            yield ('Question Commit Model deleted', 1)
+
+    @staticmethod
+    def reduce(key, values):
+        yield (key, sum(ast.literal_eval(v) for v in values))
+
+
+class RegenerateQuestionCommitAndSnapshotOneOffJob(
+        jobs.BaseMapReduceOneOffJobManager):
+    """One-off job to regenerate question commit log and snapshot models if they
+    are missing."""
+
+    @classmethod
+    def entity_classes_to_map_over(cls):
+        return [question_models.QuestionModel]
+
+    @staticmethod
+    def map(question_model):
+        if question_model.deleted:
+            return
+
+        for version in python_utils.RANGE(1, question_model.version + 1):
+            commit_log_model = (
+                question_models.QuestionCommitLogEntryModel.get_by_id(
+                    'question-%s-%s' % (question_model.id, version)))
+            if commit_log_model is None:
+                commit_log_model = (
+                    question_models.QuestionCommitLogEntryModel.create(
+                        question_model.id, version, feconf.SYSTEM_COMMITTER_ID,
+                        'create', 'Regenerate Missing Commit Log Model',
```

Comment:
This is actually kind of concerning -- these logs shouldn't be missing in the first place, and I think the way you're creating them here doesn't match the way they would naturally be created. E.g. the commit message would typically be different and more specific. There would also be specific commit_cmds which are not replicated here and that may cause problems if code logic tried to use the commit_cmds in the future.

Any insight on why/when the "missing"-ness happened? I think we need to understand that properly before we can commit to this fix.

Reply:
Hmm, no. I looked into the code but did not find any case where a commit log or snapshot model is deleted directly. This issue is seen in [two specific questions](https://docs.google.com/spreadsheets/d/1P9JnvlG_Yfed7cZtR5OcVDZ4wddIxxA9ftnAnVjsRI4/edit#gid=1965725978&range=9:9). Is it possible to directly check when these questions were created?

Reply:
Removed - job isn't required as per offline discussion.